ssl/statem/extensions.c

```diff
@@ -929,9 +929,27 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
         ret = s->session_ctx->ext.servername_cb(s, &altmp,
                                        s->session_ctx->ext.servername_arg);
 
-    if (!sent) {
-        OPENSSL_free(s->session->ext.hostname);
-        s->session->ext.hostname = NULL;
+    /*
+     * For servers, propagate the SNI hostname from the temporary
+     * storage in the SSL to the persistent SSL_SESSION, now that we
+     * know we accepted it.
+     * Clients make this copy when parsing the server's response to
+     * the extension, which is when they find out that the negotiation
+     * was successful.
+     */
+    if (s->server) {
+        if (!sent) {
+            OPENSSL_free(s->ext.hostname);
+            s->ext.hostname = NULL;
+        } else if (!s->hit || SSL_IS_TLS13(s)) {
+            /* Only store the hostname in the session if we accepted it. */
+            OPENSSL_free(s->session->ext.hostname);
+            s->session->ext.hostname = OPENSSL_strdup(s->ext.hostname);
```

Comment:
In think that in case of TLS 1.3 resumption,
when final_server_name() is called at the server side, then s->ext.hostname is currently NULL.
but s->session->ext.hostname may hold a value.
That is why we should consider making s->ext.hostname to hold the hostname at tls_parse_ctos_server_name,
even if s->hit is set.